### ISSUE TITLE
Implementation of Nearby Share 

### ISSUE THREAD

**Issue by eredrolo** [microg/android_packages_apps_GmsCore#1272]
Is it possible to add [Nearby Share](https://blog.google/products/android/nearby-share/) to MicroG and also make it work with the normal Play Services?

**Comment by fareszr** [comment]
i really hope @mar-v-in adds support for nearby share.
its really useful and a genuine competitor for airdrop, unlike previous attempts its a universal standard.

**Comment by pr0gr8mm3r** [comment]
It seems like there is an implementation for chrome under development: https://source.chromium.org/chromium/chromium/src/+/master:chrome/browser/nearby_sharing/

**Comment by samyak039** [comment]
would love to see this implemented.

**Comment by TheLastGimbus** [comment]
I'm getting more and more interested in Nearby Sharing - and also other Nearby APIs from google - they look really cool aka something that google would deprecated after a year - but they actually didn't! 

[Discord, for example, uses it to let you discover people nearby](https://support.discord.com/hc/en-us/articles/360014894392--Mobile-Nearby-Friends) - this is *such an obvious* feature that absolutley every messaging app should have, but they don't :thinking: 

One time I wanted to try it out, but I couldn't, because microG didn't have it :disappointed:  - we still had to slowly exchange usernames with my friend

I don't know why corona warnings are named `nearby` - but we should probably change this :confused: 

Also, looks like google is slowly open sourcing it:
- https://github.com/google/nearby-connections/
- https://github.com/chromium/chromium/tree/master/third_party/nearby
- https://github.com/chromium/chromium/tree/master/chrome/browser/nearby_sharing
- https://stackoverflow.com/questions/63991333/how-is-googles-nearby-share-implemented

Can we open a bounty for this? @mar-v-in, this is *THE* feature we need :pray: :eyes: :rocket:

**Comment by Relected** [comment]
I would Absolutly love it

**Comment by fbievan** [comment]
Nearby share would be an incredibly useful feature... As android beam is uh discontinued... Introducing this I think would take a lot of work... I'm gonna go mess around myself... Likely not gonna get anywhere... But still would be fun to mess around with. Willing to put up a bounty myself. But not ready to yet. Anyway I'll be back to this thread later.

**Comment by alex9099** [comment]
I'm trying to document the protocol here https://github.com/alex9099/SimpleNearby

Also I'm doing a PoC implementation in python, should then be ported to java/kotlin to use on android/microg but to do quick prototypes I find python quite simple :D 

Even though google released the code for it, there doesn't seem to be proper documentation which makes understanding the whole thing quite complicated

**Comment by Idcrafter** [comment]
nearby share is a quite useful feature which most android users have never heard of. it would be excellent if it also could be done with privacy or at least more privacy as google's implementation

**Comment by TGRush** [comment]
Nearby Share is quite the nice feature, which people without Google Play services sadly can't use, just like many other features that GMS provides. It would be great if someone could step up and start working on an implementation, especially as it would probably be more private.

**Comment by Idcrafter** [comment]
yes, i meant what if there is a Nearbyshare like feature for microG, it would be a great addition to the project, i am not good enough in programming to create such a feature.

**Comment by bblacher** [comment]
https://github.com/grishka/NearDrop/blob/master/PROTOCOL.md

Some more documentation on the protocol, may help implementing this feature.

Also, there is an implementation for MacOS:
https://github.com/grishka/NearDrop

**Comment by alex9099** [comment]
I should take a look into that project, i was following the BT/BLE approach, they are doing the TCP (so you are already in the same network) approach.
Haven't had much time (or motivation) to develop it further

**Comment by Idcrafter** [comment]
i know that NearbyShare uses a BLE Beacon to open up a Wifi-direct connection and if the phones both have network and get seperated will it switch to sending it over the internet if allowed in the settings.

**Comment by rorlig** [comment]
any updates ? is the support for nearby connections planned?

**Comment by Blisterexe** [comment]
just wanted to add that now it is called quick share, and there is an implementation for linux that may be useful to somebody who wants to implement it.
https://github.com/Martichou/rquickshare

**Comment by Idcrafter** [comment]
as far as i could gather is that all of them only work over the same network and only if there is mDNS allowed/working and some also do send Bluetooth packets to get the the play services to broadcast a mDNS signal to be findable by them.
if there is a larger afford to make the Bluetooth side work would make it usable to be considered to be implemented into microG.
microG could initate a hotspot for the time of the transfer and send over random credentials for it to have a shared network everywhere, quick-share/nearby-share supports sharing and receiving wifi credentials to then be able to establish that connection to make it work everywhere even if there is no wifi nearby and this then not only between microG to microG but also microG to play services or play services to microG?

 but i am not a good enough dev to know how this could look like or be made so is this only a speculative suggestion if this feature is planned to be added in the future.

**Comment by cyrneko** [comment]
Nearby Share, AirDrop etc. all work over local wireless networking akin to AWDL or Wi-Fi Aware (AKA Neighbour Awareness Networking)

it's definitely a larger project to take on, but it isn't simply local networking

**Comment by Idcrafter** [comment]
yeah but for the current stuff would a hotspot be a short term solution untill the reverse engineered solutions can do more then the simple lan based protocol that currently only works

**Comment by Aboudeh2010** [comment]
How is the project doing and is there any big progress

**Comment by Idcrafter** [comment]
if someone would port the reverse engineered implementations of Quick Share like what Packet is using to Android and add it into microG then would it work in the same Wifi Network but there was still nothing that happened in this space.
if someone put a bounty out for this then would this be worked on more or faster.

**Comment by Scupake** [comment]
There is this fork of NearDrop that ports it to Android: https://github.com/leonboe1/QuickDrop
I haven't tested it out, but some people online say it works great without google play services, you just have to have it running when someone tries to share something to you

**Comment by rugk** [comment]
Thanks @leonboe1 that looks quite useful! BTW I've created an issue for an F-Droid release there (https://github.com/leonboe1/QuickDrop/issues/5) that, I guess, could help many users (being subscribed) here.

**Comment by 7jari** [comment]
Samsung's Quick Share app depends on GMS for sharing to non-Galaxy devices, and I couldn't log in to my Google account there as it seems to use a different login API.
(I'm one of the very few people using stock One UI with signature spoofing enabled + microG in place of GMS lol)
